### DISCUSSION TITLE
Libdovi error with Vapoursynth

### DISCUSSION THREAD

**Discussion by Amaraldo** [quietvoid/dovi_tool#235] [General]
The dovi_tool is a great tool. I am trying to create lossless FEL reconstructed Dolby Vision screenshots using dovi_tool and Vapoursynth but I keep getting this error "No entry point found in /opt/homebrew/lib/vapoursynth/libdovi.3.1.2.dylib" when running Vapoursynth scripts.

Is this related to the fact I'm using an m1 mac? I'd appreciate any help or workaround.

**Comment by quietvoid** [comment]
Why do you have libdovi installed in the VS plugins?
It shouldn't be there.

**Comment by Amaraldo** [comment]
Ah. My mistake. Where should it be? 

Also, is it possible to reconstruct the FEL with lossless re-encoding for short spliced Dolby content? I am trying to create a database comparing the quality of standard HDR10 screenshots with their Dolby Vision counterparts. Lossy re-encoding defeats the purpose as any gains afforded by the FEL are lost in the re-encode. Thanks again!

**Reply by quietvoid** [reply]
> Ah. My mistake. Where should it be?

I don't know, where the libs go.
You're supposed to build `vs-placebo` and `vs-nlq` afterwards as they link to libdovi.
I don't provide support for this.

> Also, is it possible to reconstruct the FEL with lossless re-encoding for short spliced Dolby content

x265 technically supports 12 bit, so try it out I guess.

**Reply by Amaraldo** [reply]
I had them built and placed in the vapoursynth plugins folder, I just mistakenly put libdovi there too. 

It's not even the bit depth that I'm after, it's correctly combining the BL + FEL losslessly. Thanks for the help!

**Reply by quietvoid** [reply]
You can simply use `awsmfunc.MapDolbyVision` in VS.
https://github.com/OpusGang/awsmfunc/blob/master/src/awsmfunc/base.py#L1905

**Reply by Amaraldo** [reply]
Sorry to keep bugging you but I cannot figure out what I'm doing wrong. 
I keep getting this error:
"Error: Failed to retrieve frame 0 with error: placebo.Tonemap: Clip is missing `DolbyVisionRPU` prop for Dolby Vision mapping!"

The base_layer is a short 3 second crop of a UHD (.mkv but I tried .hevc too). Mediainfo shows profile 7 with BL+EL+RPU. I used dovi_tool to extract the EL and set that to the enhancement_layer.

**Reply by quietvoid** [reply]
Sounds like your source fillter doesn't support the `DolbyVisionRPU` frame prop.
You need to build the git version of ffms2 or l-smash.

Both clips need to have the RPU as well.

**Reply by Amaraldo** [reply]
Does the awsmfunc.MapDolbyVision function work differently to this script?

> import vapoursynth as vs
> from vapoursynth import core
> from pathlib import Path
> 
> bl_path = Path("")
> bl = core.ffms2.Source(bl_path)
> bl = bl.resize.Spline36(format=vs.YUV420P16)
> 
> bl_w = bl.width
> bl_h = bl.height
> 
> el_path = Path("")
> el = core.ffms2.Source(el_path)
> el = core.resize.Point(el, width=bl_w, height=bl_h, format=vs.YUV420P10)
> 
> hdr = core.vsnlq.MapNLQ(bl, el, "RPU.bin")
> 
> hdr.set_output()

The awsmfunc.MapDolbyVision returns shader errors to do with _15_tmp, _16_tmp issues but the script above works fine. If they both produce the same output, I can just use the script above.

**Reply by quietvoid** [reply]
Your script is wrong as it's missing the `placebo.Tonemap` filter.
It won't produce the correct output.

If you have errors in `vs-placebo`, open an issue there with more details.
It should work fine so maybe it has something to do with your setup/drivers.
